<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <style>
    #vis.vega-embed {
      width: 100%;
      display: flex;
    }

    #vis.vega-embed details,
    #vis.vega-embed details summary {
      position: relative;
    }
  </style>
  <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/vega@5"></script>
  <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/vega-lite@5.20.1"></script>
  <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/vega-embed@6"></script>
</head>
<body>
  <div id="vis"></div>
  <script>
    (function(vegaEmbed) {
      var spec = {"config": {"view": {"continuousWidth": 300, "continuousHeight": 300, "strokeWidth": 0}, "legend": {"labelLimit": 0}, "mark": {"stroke": "white", "strokeOpacity": 1.0, "strokeWidth": 1.0}, "title": {"fontSize": 15, "offset": 10}}, "layer": [{"layer": [{"mark": {"type": "arc", "outerRadius": 190}, "encoding": {"color": {"field": "L_DESCRIPTION", "scale": {"domain": ["J : Translation, ribosomal structure and biogenesis", "G : Carbohydrate transport and metabolism", "L : Replication, recombination and repair", "E : Amino acid transport and metabolism", "H : Coenzyme transport and metabolism", "F : Nucleotide transport and metabolism", "D : Cell cycle control, cell division, chromosome partitioning", "O : Posttranslational modification, protein turnover, chaperones", "M : Cell wall/membrane/envelope biogenesis", "C : Energy production and conversion", "P : Inorganic ion transport and metabolism", "K : Transcription", "V : Defense mechanisms", "I : Lipid transport and metabolism", "S : Function unknown", "T : Signal transduction mechanisms", "U : Intracellular trafficking, secretion, and vesicular transport", "R : General function prediction only", "A : RNA processing and modification", "B : Chromatin structure and dynamics", "N : Cell motility"], "range": ["#FCCCFC", "#CCFCFC", "#FCDCDC", "#DCFCFC", "#DCDCFC", "#DCECFC", "#FCFCCC", "#9CFC9C", "#DCFCAC", "#BCFCFC", "#CCCCFC", "#FCDCEC", "#FCFCAC", "#DCCCFC", "#CCCCCC", "#ECFCAC", "#9CFCAC", "#E0E0E0", "#FCDCFC", "#FCDCCC", "#CCFCAC"]}, "title": "", "type": "nominal"}, "order": {"field": "COUNT", "sort": "descending", "type": "quantitative"}, "theta": {"field": "COUNT", "stack": true, "type": "quantitative"}, "tooltip": [{"field": "DESCRIPTION", "type": "nominal"}, {"field": "LETTER", "type": "nominal"}, {"field": "COUNT", "type": "quantitative"}, {"field": "RATIO(%)", "type": "nominal"}]}, "title": "COG Functional Classification"}, {"mark": {"type": "text", "radius": 175, "size": 10, "stroke": "black", "strokeOpacity": 1.0, "strokeWidth": 1.0}, "encoding": {"color": {"field": "L_DESCRIPTION", "scale": {"domain": ["J : Translation, ribosomal structure and biogenesis", "G : Carbohydrate transport and metabolism", "L : Replication, recombination and repair", "E : Amino acid transport and metabolism", "H : Coenzyme transport and metabolism", "F : Nucleotide transport and metabolism", "D : Cell cycle control, cell division, chromosome partitioning", "O : Posttranslational modification, protein turnover, chaperones", "M : Cell wall/membrane/envelope biogenesis", "C : Energy production and conversion", "P : Inorganic ion transport and metabolism", "K : Transcription", "V : Defense mechanisms", "I : Lipid transport and metabolism", "S : Function unknown", "T : Signal transduction mechanisms", "U : Intracellular trafficking, secretion, and vesicular transport", "R : General function prediction only", "A : RNA processing and modification", "B : Chromatin structure and dynamics", "N : Cell motility"], "range": ["#FCCCFC", "#CCFCFC", "#FCDCDC", "#DCFCFC", "#DCDCFC", "#DCECFC", "#FCFCCC", "#9CFC9C", "#DCFCAC", "#BCFCFC", "#CCCCFC", "#FCDCEC", "#FCFCAC", "#DCCCFC", "#CCCCCC", "#ECFCAC", "#9CFCAC", "#E0E0E0", "#FCDCFC", "#FCDCCC", "#CCFCAC"]}, "title": "", "type": "nominal"}, "order": {"field": "COUNT", "sort": "descending", "type": "quantitative"}, "text": {"field": "VISIBLE_LETTER", "type": "nominal"}, "theta": {"field": "COUNT", "stack": true, "type": "quantitative"}, "tooltip": [{"field": "DESCRIPTION", "type": "nominal"}, {"field": "LETTER", "type": "nominal"}, {"field": "COUNT", "type": "quantitative"}, {"field": "RATIO(%)", "type": "nominal"}]}, "title": "COG Functional Classification"}]}], "data": {"name": "data-8d7e668176f66e5c37ccae077c95651b"}, "height": 380, "width": 380, "$schema": "https://vega.github.io/schema/vega-lite/v5.20.1.json", "datasets": {"data-8d7e668176f66e5c37ccae077c95651b": [{"LETTER": "J", "COUNT": 123, "GROUP": "INFORMATION STORAGE AND PROCESSING", "COLOR": "#FCCCFC", "DESCRIPTION": "Translation, ribosomal structure and biogenesis", "RATIO": 26.68112798264642, "RATIO(%)": "26.68%", "L_DESCRIPTION": "J : Translation, ribosomal structure and biogenesis", "VISIBLE_LETTER": "J"}, {"LETTER": "G", "COUNT": 43, "GROUP": "METABOLISM", "COLOR": "#CCFCFC", "DESCRIPTION": "Carbohydrate transport and metabolism", "RATIO": 9.327548806941431, "RATIO(%)": "9.33%", "L_DESCRIPTION": "G : Carbohydrate transport and metabolism", "VISIBLE_LETTER": "G"}, {"LETTER": "L", "COUNT": 38, "GROUP": "INFORMATION STORAGE AND PROCESSING", "COLOR": "#FCDCDC", "DESCRIPTION": "Replication, recombination and repair", "RATIO": 8.24295010845987, "RATIO(%)": "8.24%", "L_DESCRIPTION": "L : Replication, recombination and repair", "VISIBLE_LETTER": "L"}, {"LETTER": "E", "COUNT": 25, "GROUP": "METABOLISM", "COLOR": "#DCFCFC", "DESCRIPTION": "Amino acid transport and metabolism", "RATIO": 5.42299349240781, "RATIO(%)": "5.42%", "L_DESCRIPTION": "E : Amino acid transport and metabolism", "VISIBLE_LETTER": "E"}, {"LETTER": "H", "COUNT": 24, "GROUP": "METABOLISM", "COLOR": "#DCDCFC", "DESCRIPTION": "Coenzyme transport and metabolism", "RATIO": 5.206073752711497, "RATIO(%)": "5.21%", "L_DESCRIPTION": "H : Coenzyme transport and metabolism", "VISIBLE_LETTER": "H"}, {"LETTER": "F", "COUNT": 23, "GROUP": "METABOLISM", "COLOR": "#DCECFC", "DESCRIPTION": "Nucleotide transport and metabolism", "RATIO": 4.989154013015185, "RATIO(%)": "4.99%", "L_DESCRIPTION": "F : Nucleotide transport and metabolism", "VISIBLE_LETTER": "F"}, {"LETTER": "D", "COUNT": 22, "GROUP": "CELLULAR PROCESSES AND SIGNALING", "COLOR": "#FCFCCC", "DESCRIPTION": "Cell cycle control, cell division, chromosome partitioning", "RATIO": 4.772234273318872, "RATIO(%)": "4.77%", "L_DESCRIPTION": "D : Cell cycle control, cell division, chromosome partitioning", "VISIBLE_LETTER": "D"}, {"LETTER": "O", "COUNT": 21, "GROUP": "CELLULAR PROCESSES AND SIGNALING", "COLOR": "#9CFC9C", "DESCRIPTION": "Posttranslational modification, protein turnover, chaperones", "RATIO": 4.55531453362256, "RATIO(%)": "4.56%", "L_DESCRIPTION": "O : Posttranslational modification, protein turnover, chaperones", "VISIBLE_LETTER": "O"}, {"LETTER": "M", "COUNT": 21, "GROUP": "CELLULAR PROCESSES AND SIGNALING", "COLOR": "#DCFCAC", "DESCRIPTION": "Cell wall/membrane/envelope biogenesis", "RATIO": 4.55531453362256, "RATIO(%)": "4.56%", "L_DESCRIPTION": "M : Cell wall/membrane/envelope biogenesis", "VISIBLE_LETTER": "M"}, {"LETTER": "C", "COUNT": 20, "GROUP": "METABOLISM", "COLOR": "#BCFCFC", "DESCRIPTION": "Energy production and conversion", "RATIO": 4.3383947939262475, "RATIO(%)": "4.34%", "L_DESCRIPTION": "C : Energy production and conversion", "VISIBLE_LETTER": "C"}, {"LETTER": "P", "COUNT": 18, "GROUP": "METABOLISM", "COLOR": "#CCCCFC", "DESCRIPTION": "Inorganic ion transport and metabolism", "RATIO": 3.9045553145336225, "RATIO(%)": "3.90%", "L_DESCRIPTION": "P : Inorganic ion transport and metabolism", "VISIBLE_LETTER": "P"}, {"LETTER": "K", "COUNT": 18, "GROUP": "INFORMATION STORAGE AND PROCESSING", "COLOR": "#FCDCEC", "DESCRIPTION": "Transcription", "RATIO": 3.9045553145336225, "RATIO(%)": "3.90%", "L_DESCRIPTION": "K : Transcription", "VISIBLE_LETTER": "K"}, {"LETTER": "V", "COUNT": 16, "GROUP": "CELLULAR PROCESSES AND SIGNALING", "COLOR": "#FCFCAC", "DESCRIPTION": "Defense mechanisms", "RATIO": 3.470715835140998, "RATIO(%)": "3.47%", "L_DESCRIPTION": "V : Defense mechanisms", "VISIBLE_LETTER": "V"}, {"LETTER": "I", "COUNT": 13, "GROUP": "METABOLISM", "COLOR": "#DCCCFC", "DESCRIPTION": "Lipid transport and metabolism", "RATIO": 2.8199566160520604, "RATIO(%)": "2.82%", "L_DESCRIPTION": "I : Lipid transport and metabolism", "VISIBLE_LETTER": "I"}, {"LETTER": "S", "COUNT": 12, "GROUP": "POORLY CHARACTERIZED", "COLOR": "#CCCCCC", "DESCRIPTION": "Function unknown", "RATIO": 2.6030368763557483, "RATIO(%)": "2.60%", "L_DESCRIPTION": "S : Function unknown", "VISIBLE_LETTER": "S"}, {"LETTER": "T", "COUNT": 8, "GROUP": "CELLULAR PROCESSES AND SIGNALING", "COLOR": "#ECFCAC", "DESCRIPTION": "Signal transduction mechanisms", "RATIO": 1.735357917570499, "RATIO(%)": "1.74%", "L_DESCRIPTION": "T : Signal transduction mechanisms", "VISIBLE_LETTER": "T"}, {"LETTER": "U", "COUNT": 6, "GROUP": "CELLULAR PROCESSES AND SIGNALING", "COLOR": "#9CFCAC", "DESCRIPTION": "Intracellular trafficking, secretion, and vesicular transport", "RATIO": 1.3015184381778742, "RATIO(%)": "1.30%", "L_DESCRIPTION": "U : Intracellular trafficking, secretion, and vesicular transport", "VISIBLE_LETTER": "U"}, {"LETTER": "R", "COUNT": 5, "GROUP": "POORLY CHARACTERIZED", "COLOR": "#E0E0E0", "DESCRIPTION": "General function prediction only", "RATIO": 1.0845986984815619, "RATIO(%)": "1.08%", "L_DESCRIPTION": "R : General function prediction only", "VISIBLE_LETTER": "R"}, {"LETTER": "A", "COUNT": 3, "GROUP": "INFORMATION STORAGE AND PROCESSING", "COLOR": "#FCDCFC", "DESCRIPTION": "RNA processing and modification", "RATIO": 0.6507592190889371, "RATIO(%)": "0.65%", "L_DESCRIPTION": "A : RNA processing and modification", "VISIBLE_LETTER": ""}, {"LETTER": "B", "COUNT": 1, "GROUP": "INFORMATION STORAGE AND PROCESSING", "COLOR": "#FCDCCC", "DESCRIPTION": "Chromatin structure and dynamics", "RATIO": 0.21691973969631237, "RATIO(%)": "0.22%", "L_DESCRIPTION": "B : Chromatin structure and dynamics", "VISIBLE_LETTER": ""}, {"LETTER": "N", "COUNT": 1, "GROUP": "CELLULAR PROCESSES AND SIGNALING", "COLOR": "#CCFCAC", "DESCRIPTION": "Cell motility", "RATIO": 0.21691973969631237, "RATIO(%)": "0.22%", "L_DESCRIPTION": "N : Cell motility", "VISIBLE_LETTER": ""}]}};
      var embedOpt = {"mode": "vega-lite"};

      function showError(el, error){
          el.innerHTML = ('<div style="color:red;">'
                          + '<p>JavaScript Error: ' + error.message + '</p>'
                          + "<p>This usually means there's a typo in your chart specification. "
                          + "See the javascript console for the full traceback.</p>"
                          + '</div>');
          throw error;
      }
      const el = document.getElementById('vis');
      vegaEmbed("#vis", spec, embedOpt)
        .catch(error => showError(el, error));
    })(vegaEmbed);

  </script>
</body>
</html>